group/assignment6/assignment6.py

```diff
@@ -0,0 +1,112 @@
+    to the current lot.
+
+    :param start_state: order of cars in the start of the rearrangement
+    :param end_state: order of cars after rearrangement
+    :yields: move steps. Each move is represented as a tuple with two indeces,
+             the first index is the number of lot from which we move the car,
+             the second index is the number of lot to which we move the car
+    '''
+    if len(start_state) != len(end_state):
+        raise IndexError('The start state and end state have different lengths.')
+    current_state = start_state.copy()
+    for lot in range(len(current_state)):
+        right_car = end_state[lot]
+        current_right_car = find_car_lot(current_state, right_car)
+        if not current_state[lot] == 0:
+            empty = find_empty(current_state)
+            current_state = move_to_lot(current_state, lot, empty)
+            yield (lot, empty)
+        current_state = move_to_lot(current_state, current_right_car, lot)
+        yield (current_right_car, lot)
+        if current_state == end_state:
+            break    
+    
+def compute_efficient_moves(start_state, end_state):
+    '''
+    More efficiently omputes a sequence of moves that are required to rearrange cars
+    from the given start state to the end state.
+    We look up which lot is empty and which car should be in this slot.
+    Then we move the required car to the empty lot and repeat the procedure.
+
+    :param start_state: order of cars in the start of the rearrangement
+    :param end_state: order of cars after rearrangement
+    :yields: move steps. Each move is represented as a tuple with two indeces,
+             the 1st index is the number of lot from which we move the car,
+             the 2nd index is the number of lot to which we move the car
+    '''
+    if len(start_state) != len(end_state):
+        raise IndexError('The start state and end state have different lengths.')
+    current_state = start_state.copy()
+    while current_state != end_state:
+        empty = find_empty(current_state)
+        right_car = end_state[empty]
+        current_right_car = find_car_lot(current_state, right_car)
+        current_state = move_to_lot(current_state, current_right_car, empty)
+        yield (current_right_car, empty)
+    
+def apply_moves(start_state, moves):
+    '''
+    Function for checking the correctness of our results. We take 
+    the start state and apply the move steps.
+    
+    :param start_state: order of cars in the start of the rearrangement
+    :param moves: generator, sequence of moves that we need to apply.
+                  Each move is represented as a tuple with two indeces, 
+                  the 1st index is the number of lot from which we move the car,
+                  the 2nd index is the number of lot to which we move the car
+    :returns: end state, list of car numbers. 
+    '''
+    current_state = start_state.copy()
```

Comment:
are you sure that we need to copy start_state? can you explain it?

Reply:
You mean here or everywhere? I copy it because I don't want to modify start_state to keep it for future reference. But as we don't use the start_state, maybe it's ok to modify it.

Reply:
I think that we don''t use the start_state so maybe you can try not to copy it and run the tests to check if everything is ok